### ISSUE TITLE
Missing logentries

### ISSUE THREAD

**Issue by Bizzelicious** [sivasamyk/logtrail#32]
Hi!
First of all, thank you for this wonderful plugin. We currently ship logs to both ES and Papertrail via Logstash to be able to both tail and analyze our logs. But we hope to change that to ES only with this plugin.

However, I've noticed that LogTrail seems to drop/not show some logs.

In papertrail we have the following logentries:
```
Bamboo-host bamboo:  2017-01-12 16:00:30,966 WARN [ActiveMQ Transport: tcp:///xxxxx:51553@54663] [Transport] Transport Connection to: tcp://xxxxx:51553 failed: java.io.EOFException
Bamboo-host bamboo:  2017-01-12 16:00:34,472 INFO [http-nio-8085-exec-15] [AccessLogFilter] xxxxx GET https://xxxxx/userlogin!doDefault.action 233579kb
Bamboo-host bamboo:  2017-01-12 16:00:34,778 INFO [http-nio-8085-exec-12] [DefaultAuthenticator] login : '' could not be authenticated with the given password
Bamboo-host bamboo:  2017-01-12 16:00:34,778 WARN [http-nio-8085-exec-12] [DefaultAuthenticator] login : '' tried to login but they do not have USE permission or weren't found. Deleting remember me cookie.
Bamboo-host bamboo:  2017-01-12 16:00:34,778 INFO [http-nio-8085-exec-12] [BambooAuthenticator] Finished doing authentication checks for , took 744.3 ?s
Bamboo-host bamboo:  2017-01-12 16:00:34,778 INFO [http-nio-8085-exec-12] [AccessLogFilter] xxxxx POST https://xxxxx/userlogin.action 226212kb
Bamboo-host bamboo:  2017-01-12 16:00:35,180 INFO [http-nio-8085-exec-13] [AccessLogFilter] xxxxx GET https://xxxxx/rest/pref/latest/user/?expand=preferences.preference&_=1484236794961 219098kb
Bamboo-host bamboo:  2017-01-12 16:00:35,240 INFO [http-nio-8085-exec-20] [AccessLogFilter] xxxxx GET https://xxxxx/rest/menu/latest/appswitcher?_=1484236794962 218850kb
```
In LogTrail we have the following logentries:
```
Bamboo-host bamboo:  2017-01-12 16:00:30,966 WARN [ActiveMQ Transport: tcp:///xxxxx:51553@54663] [Transport] Transport Connection to: tcp://xxxxx:51553 failed: java.io.EOFException
Bamboo-host bamboo:  2017-01-12 16:00:34,778 INFO [http-nio-8085-exec-12] [BambooAuthenticator] Finished doing authentication checks for , took 744.3 ?s
Bamboo-host bamboo:  2017-01-12 16:00:34,778 INFO [http-nio-8085-exec-12] [DefaultAuthenticator] login : '' could not be authenticated with the given password
Bamboo-host bamboo:  2017-01-12 16:00:35,240 INFO [http-nio-8085-exec-20] [AccessLogFilter] xxxxx GET https://xxxxxx/rest/menu/latest/appswitcher?_=1484236794962 218850kb
```

As you can see, the first and last logs are the same, but the logs in between are different, where Papertrail is "the right one". I can see all the logs in Kibana, but LogTrail seems to miss them. 
 We use the logtrail-4.x-0.1.5 version and unfortunately we're stuck with Kibana 4.5 until AWS updates their Elasticsearch service. 

Our Logtrail conf:
```
  "tail_interval_in_seconds": 3,
  "max_buckets": 500,
  "nested_objects" : false,
  "default_time_range_in_days" : 0,
```

I have a feeling that this problem is due too our low refresh rate and when we refresh, aka fetch new logs from ES, the logs we're missing are not quite ready in ES yet, like they're still being indexed or similar. And I guess that Logtrail only checks for logs with a higher timestamp than the last Logtrail entry every time it's refreshing for new logs? 
Afaik Kibana for example does not do this, Kibana refreshes the whole search interval everytime, so even if a document took a longer time to index, it would show up the next refresh.

**Comment by C0rn3j** [comment]
>And I guess that Logtrail only checks for logs with a higher timestamp than the last Logtrail entry every time it's refreshing for new logs?

I just wanted to create a separate issue but you were faster ^^

https://github.com/sivasamyk/logtrail/issues/31#issuecomment-272207817

**Comment by sivasamyk** [comment]
@Bizzelicious Your guess about logtrail tail logic is correct. Its checks for all logs whose timestamp is less than and equal to last event and then eliminate duplicates. I assume you use clustered elasticsearch setup. If ES receives events A,B,C,D,E,F in this order, when logtrail queries for all events whose timestamp is >= B's TS, how come ES returns E,F before C,D. You are assuming C & D are still being indexed? 

Anyways thanks for this interesting bug. Let me analyze more on this.

**Comment by Bizzelicious** [comment]
@C0rn3j Yes, if a entry is missing and I reload the page, the missing entry will re-appear.

@sivasamyk Yes, we are using a clustred ES setup. 
About the events, I'm not an ES expert so "indexing" was just an assumption,. My point was that some events might take longer to "get ready" for fetching.

I guess a solution is to make Logtrail query ES for last TS minus 5 seconds or similar. This could be added as a config.

**Comment by C0rn3j** [comment]
@Bizzelicious Imagine you have a misconfigured time on a device or perhaps even a device in a different timezone, that device is 1 hour ahead.

Getting just one log from the device would block all the other logs for 1 hour, and there would be more logs, so you'd still only ever see that one device, which is not much different from how LogTrail works right now.

I think that a workable solution could be simply showing logs as they arrive and not caring too much about time of the last received log

**Comment by Bizzelicious** [comment]
@C0rn3j That's very true! Just like mentioned in #31 and my suggestion would not solve that.

**Comment by sivasamyk** [comment]
Logtrail by default uses the timestamp field added by logstash ( this field is added by logstash based on time it receives the event) to order the events and fetch during tail. Irrespective of the timezone the device operates in, this timestamp will be in UTC. Based on this I think the 5 seconds logic proposed by @Bizzelicious should work.

**Comment by Bizzelicious** [comment]
@sivasamyk I guess you mean the @timestamp field right? It's not entirely true that the @timestamp field is based on when logstash receives the log. It is that by default, but a lot of people are using grok filters to parse out the "real" timestamp from the log and replace @timestamp with that value via the Date plugin (https://www.elastic.co/guide/en/logstash/current/plugins-filters-date.html). 

So a problem with going with my suggestion is that if we're running a clustered application logging to Logstash which are changing the @timestamp to the "log timestamp" as explained above and one of the cluster nodes are falling behind in outputting the log to logstash (this could be due to backpressure if we're running beats for example), or if NTP gets out of sync on the node. We get the same problem as discussed in issue #31 and as @C0rn3j mentioned earlier. 
The problem is that the application on the node which is OK will log as usual and LogTrail will use those logs as the time baseline. This means if the second node is falling behind more than 5 seconds, the logs from that node will never reach LogTrail since those logs would be seen as "too old".

On a second thought, this might not be a problem if we can change the "timestamp" that LogTrail uses default. We, for example are using a field called "inputTime" to get the time logstash received the event (instead of @timestamp, which in our case is the real log time). So we could use that field instead.

**Comment by sivasamyk** [comment]
@Bizzelicious Agreed. You can use a different field to log_received_time added by logstash and map it to timestamp in logtrail.json. In this case -5 second logic should work. I might need your help in testing the fix once it is ready.

**Comment by Bizzelicious** [comment]
@sivasamyk That's no problem at all, I have a test environment ready

**Comment by sivasamyk** [comment]
@Bizzelicious I have added new field in logtrail.json es_index_margin_in_seconds. Use the attached plugin and let me know if works. If you required, you can change the margin to 10 seconds.
[logtrail-4.x-0.1.6.tar.gz](https://github.com/sivasamyk/logtrail/files/708216/logtrail-4.x-0.1.6.tar.gz)

**Comment by sivasamyk** [comment]
@Bizzelicious @C0rn3j  Do you have multiple elasticsearch hosts configured in your elasticsearch output of logstash? Can you please check if the issue is reproducible if you have only 1 host configured in es output of logstash (you can test this scenario without the above fix)? When we give multiple hosts is es output of logstash it will load balance among the hosts.

**Comment by C0rn3j** [comment]
Don't have test env right now(and likely won't have time in the near future to set one up), but yes I was using 5 hosts.

**Comment by Bizzelicious** [comment]
Hi!
I can confirm that the setting is working. However, now we get a lot of duplicated logs and alot of the logs are out of place. Example: "Username4 log" has a time which should be before the "Username1 log", but this might be a problem because I use a low refresh rate since LogTrail only "pastes" the next log it sees and won't put it in the right place.

```
**Bamboo-host bamboo:  2017-01-17 11:03:29,361 INFO [http-nio-8085-exec-63] [AccessLogFilter] Username1 GET https://my.url.here.com/rest/api/latest/plan/favicon/RT-AUT?_=1484634967613 217091kb**
**Bamboo-host bamboo:  2017-01-17 11:03:29,752 INFO [http-nio-8085-exec-7] [AccessLogFilter] Username2 GET https://my.url.here.com/chain/admin/ajax/getChains.action?planKey=EC-WW&_=1484648512576 200851kb**
**Bamboo-host bamboo:  2017-01-17 11:03:29,769 INFO [http-nio-8085-exec-13] [AccessLogFilter] Username3 GET https://my.url.here.com/ajax/viewPlanUpdates.action?sinceSystemTime=1484650988747&lastProject=SGIT&_=1484650673113 198970kb**
Bamboo-host bamboo:  2017-01-17 11:03:28,581 INFO [http-nio-8085-exec-15] [AccessLogFilter] Username4 GET https://my.url.here.com/ajax/viewPlanUpdates.action?sinceSystemTime=1484650987593&lastProject=&_=1484586710813 230880kb
Bamboo-host bamboo:  2017-01-17 11:03:29,705 WARN [http-nio-8085-exec-31] [AuthorizationLoggerListener] Authorization failed: org.acegisecurity.AccessDeniedException: Access is denied; authenticated principal: org.acegisecurity.providers.anonymous.AnonymousAuthenticationToken@6faaf9b0: Username: 
anonymousUser; Password: [PROTECTED]; Authenticated: true; Details: org.acegisecurity.ui.WebAuthenticationDetails@ffff8868: RemoteIpAddress: x.x.X.x; SessionId: null; Granted Authorities: ROLE_ANONYMOUS; secure object: com.atlassian.bamboo.webwork.StarterAction@18f40b1b; configuration attributes: [WW_READ, GLOBAL_READ]
Bamboo-host bamboo:  2017-01-17 11:03:29,422 INFO [http-nio-8085-exec-8] [AccessLogFilter] Username5 GET https://my.url.here.com/chain/admin/ajax/getChains.action?planKey=RT-AUT&_=1484634967614 216700kb
Bamboo-host bamboo:  2017-01-17 11:03:29,703 INFO [http-nio-8085-exec-31] [AccessLogFilter] 10.63.3.200 GET https://my.url.here.com/ 213199kb
Bamboo-host bamboo:  2017-01-17 11:03:30,771 WARN [ActiveMQ Transport: tcp:///10.63.3.200:16512@54663] [Transport] Transport Connection to: tcp://10.63.3.200:16512 failed: java.io.EOFException
**Bamboo-host bamboo:  2017-01-17 11:03:29,361 INFO [http-nio-8085-exec-63] [AccessLogFilter] Username1 GET https://my.url.here.com/rest/api/latest/plan/favicon/RT-AUT?_=1484634967613 217091kb**
Bamboo-host bamboo:  2017-01-17 11:03:29,710 INFO [http-nio-8085-exec-32] [AccessLogFilter] x.x.x.x GET https://my.url.here.com/userlogin!doDefault.action;jsessionid=28CD135D96159764DEAECA050DC3E628?os_destination=%2Fstart.action 212419kb
**Bamboo-host bamboo:  2017-01-17 11:03:29,752 INFO [http-nio-8085-exec-7] [AccessLogFilter] Username2 GET https://my.url.here.com/chain/admin/ajax/getChains.action?planKey=EC-WW&_=1484648512576 200851kb**
**Bamboo-host bamboo:  2017-01-17 11:03:29,769 INFO [http-nio-8085-exec-13] [AccessLogFilter] Username3 GET https://my.url.here.com/ajax/viewPlanUpdates.action?sinceSystemTime=1484650988747&lastProject=SGIT&_=1484650673113 198970kb**
```

In Logstash we use one endpoint which is a cluster endpoint for two ES instances which can scale out. I can probably scale this down to one host for a testing purpose. I will check that out.

**Comment by sivasamyk** [comment]
Thanks for testing the fix. The attached plugin has the fix for duplicate events issue (maintaining a backup map to check for duplicate ids). Regarding the order of events in tail, its a bit tricky to fix. This update also has some timezone related fixes. I hope it should not affect you.
[logtrail-4.x-0.1.6.tar.gz](https://github.com/sivasamyk/logtrail/files/711278/logtrail-4.x-0.1.6.tar.gz)

**Comment by Bizzelicious** [comment]
@sivasamyk Just tested the fix, I have not received any duplicates during this last test period (~30min).  Very nice!
I will test with only one ES host tonight too see how the duplicated logs (with the old version) works out and also check if we receive logs in the right order with only one host.

**Comment by sivasamyk** [comment]
👍

**Comment by Bizzelicious** [comment]
So I scaled down our cluster to only one node, no master nodes only one data node. I got the exact same problem as I do with several nodes. If I use the old version of LogTrail, I miss logentries. But the new version works just fine. However, Logs are not in order even with one host, eg:
```
Bamboo-host bamboo:  2017-01-17 23:27:22,863 INFO [http-nio-8085-exec-13] [AccessLogFilter] x.x.x.x GET https://my.url.com/userlogin!doDefault.action 193429kb
Bamboo-host bamboo:  2017-01-17 23:27:23,410 INFO [http-nio-8085-exec-26] [AccessLogFilter] x.x.x.x GET https://my.url.com/userlogin!doDefault.action;jsessionid=2224D29C2C2E34B5C7A1A26DBED8D6B7?os_destination=%2Fstart.action 185261kb
Bamboo-host bamboo:  2017-01-17 23:27:30,421 INFO [http-nio-8085-exec-63] [BambooAuthenticator] Finished doing authentication checks for , took 796.0 ?s
Bamboo-host bamboo:  2017-01-17 23:27:31,141 INFO [http-nio-8085-exec-47] [AccessLogFilter] x.x.x.x POST https://my.url.com/userlogin.action 149838kb
Bamboo-host bamboo:  2017-01-17 23:27:31,511 INFO [http-nio-8085-exec-11] [AccessLogFilter] x.x.x.x GET https://my.url.com/rest/menu/latest/appswitcher?_=1484695607883 142567kb
Bamboo-host bamboo:  2017-01-17 23:27:31,849 WARN [http-nio-8085-exec-29] [DefaultAuthenticator] login : '' tried to login but they do not have USE permission or weren't found. Deleting remember me cookie.
Bamboo-host bamboo:  2017-01-17 23:27:32,176 INFO [http-nio-8085-exec-7] [AccessLogFilter] x.x.x.x GET https://my.url.com/rest/pref/latest/user/?expand=preferences.preference&_=1484695608556 134180kb
Bamboo-host bamboo:  2017-01-17 23:27:28,903 WARN [http-nio-8085-exec-33] [DefaultAuthenticator] login : '' tried to login but they do not have USE permission or weren't found. Deleting remember me cookie.
Bamboo-host bamboo:  2017-01-17 23:27:28,903 INFO [http-nio-8085-exec-33] [DefaultAuthenticator] login : '' could not be authenticated with the given password
```

So I guess that some documents (logs) takes a little bit longer to be indexed in ES than others, we're using 4 shards per index so it might be the "load balancing" of the shards which adds some latency to the indexing. 
I guess a workaround is to use a longer "tail_interval_in_seconds", we're using 3 right now. But with high interval, we are loosing some of the "tail feeling".

**Comment by danielmotaleite** [comment]
i think that the ```tail_interval_in_seconds``` should be at least the elasticsearch index refresh time (5s default, IIRC). Lower than that, you are getting the old index list cache instead without new entries.

Having a low index refresh time hurts ES performance a lot (as it keep discarding any cache), but having everything in sync should not miss any event.

one workaround could be "fake it" ! :)
have a "buffer" for the ES index refresh rate and then print then based the event timestamp... 2 events with 1 second between then would be print with 1 second difference... "real time" events, with a few seconds (ES index refresh time) lag

this still add the lag, but will make most people happy, as the main problem of big buffers/refresh delays is big log jumps that make things harder to track. A smooth log makes the lag a minor problem :)

**Comment by Bizzelicious** [comment]
@danielmotaleite The default Index refreshtime is 1s according to: https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#dynamic-index-settings. 

Even if I set it above or at 5s as you mentioned, some logs are still not in order. However logs which are not in order is much less with a higher refresh time obviously.

You mention that the "lag"(refresh time in this case) is not a big deal, but it is. Imagine that you are testing your application and you have debugging on too see what is happening when you are pressing a button for example. It will be very hard to know which log entries are actually relevant or not, since the logs will be several seconds late.

**Comment by danielmotaleite** [comment]
you are right about the refresh... that is the default... but i think that filebeat template sets the default to 5s for filebeat-*. I suspect logstash template will do the same (i can not check this one, as i have increased it to 10s on my system)

For the lag, if the lag is "constant", one can track the errors mostly by waiting those seconds. Also most (if not all) errors should have the timestamp, to help track what events are you looking for. Of course, not all user cases can accept that lag, i imagine on a heavy log tail, the lag would make tracking very hard. that was also why i said "most user" :)

Having a "near" realtime tail on a centralize log storage that is both low latency and fast for old queries is not easy. If near realtime is really needed, maybe having syslog/logstash writing the events to a log file and rotate it every hour/day might be better :)

anyway, maybe logtrail can fetch data not since last timestamp, but since last returned timestamp for each index. If there are new logs since, they will be also fetched. This could avoid the ```es_index_time_offset_in_seconds```

For the out of order, i do not see then, If the out of orders events no not exist when using only one index, then probably  the out of order is caused by the way the data is presented from the different indexes (one by one instead of getting all of then and sort then)

**Comment by sivasamyk** [comment]
Logstash default template sets the refresh_interval to 5s.

https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json

**Comment by Bizzelicious** [comment]
@danielmotaleite You are correct about the refresh times on the templates. 

Of course you can "learn" the lag, but that's not a solution.
Having near real time tailing on a centralized logging solution is hard indeed, that's why we're currently sending our Dev and QA logs to both ES and Papertrail. ES for the long term storage, searching, analyzing etc and Papertrail for real-time tailing.

The original problem we had was solved with `es_index_time_offset_in_seconds`. We are only using Logtrail to fetch from **one Index** (with 4 shards) so what you are suggesting would not help us nor solve the original issue.

About the "out of order problem", I think what you are saying is kind of right, but instead of Index, it is the shards in the Index which is the problem (We are only fetching data from one index, remember).  I gave an example before that some shards might have a higher load than the others so, for example, the internal es "loadbalancing" in the index might add some latency to the log getting saved to a high loaded shard. Which means that that log will be indexed later than the others, hence the log falls behind in Logtrail.

> So I guess that some documents (logs) takes a little bit longer to be indexed in ES than others, we're using 4 shards per index so it might be the "load balancing" of the shards which adds some latency to the indexing.

**Comment by sivasamyk** [comment]
ES uses following formula to route the documents to shard 
`shard = hash(routing) % number_of_primary_shards`

The `routing` value is an arbitrary string, which defaults to the document’s _id but can also be set to a custom value.

https://www.elastic.co/guide/en/elasticsearch/guide/current/routing-value.html

**Comment by sivasamyk** [comment]
@Bizzelicious I can you please try having a custom template in logstash with `index.refresh_interva`l as 1s ( assuming logtrail your `tail_interval_in_seconds` is 3 seconds and `es_index_time_offset_in_seconds` is 0 seconds) and see if the issue is still reproducible.

The 5.x template for logstash can be found at https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/elasticsearch-template-es5x.json
You can set the `index.refresh_interval` to 1s ( something less that your tail interval in logtail).

In your logstash elasticsearch output conf you can set the `template` to path of this custom template.

**Comment by Bizzelicious** [comment]
Sure, ill get back to you with the results.
